### DISCUSSION TITLE
How to create new sendable versions/commits

### DISCUSSION THREAD

**Discussion by etclub** [stepwisehq/prosemirror-collab-commit#2] [General]
I get lost with the doc.
After `sendableCommit(state)`, the client should mark the current state "saved",
so that the next `sendableCommit(state)` would generate commits based on the saved one, right?

But it is not mentioned in the doc.

**Comment by ProTip** [comment]
Hi @etclub ,

`sendableCommit(state)` will keep returning the same commit until the plugin has received confirmation from the authority.

This is covered in the `README` under **Client Receive Commit** which handles the broadcast from the last step under **Authority Map Commit**. The behavior and procedure is very similar to the official `prosemirror-collab` plugin.

**Reply by etclub** [reply]
Yes, I find `sendableCommit(state)` keeps returning the same commit.
So, after a user hit a button and `sendableCommit(state)`, no matter what changes afterwords, those changes are not commitable before it receives and applies **Authority Map Commit**, right?

BTW, under **Authority Map Commit** in the `README`
```
...
const newCommits = await getDocumentCommitsSince(commitJson.version) // TODO: Implement
...
```
here should it be `newCommitJson.version`?

**Reply by ProTip** [reply]
> So, after a user hit a button and sendableCommit(state), no matter what changes afterwords, those changes are not commitable before it receives and applies Authority Map Commit, right?

Correct. Once an "in-flight" commit is created it must be processed by the authority and the confirmation received by the plugin before another commit can be created from local steps.

Further, just like with `prosemirror-collab` the plugin must receive the commits in order from the authority; it will ignore old and future commit versions. I use code that buffers commits from the authority and then applies them in order until a version gap is reached or the buffer is emptied.

> here should it be newCommitJson.version?

Yes, it should be `newCommitJson.version`.

**Reply by etclub** [reply]
Well, I did not use `prosemirror-collab` before. Forgive me if I asked stupid questions.

I have two more questions:
1. In theory, the server has to keep all the commit history to make `getDocumentCommitsSince` work, in case someone loads a very old version and commits after a very long time, right? 
```
const newCommits = await getDocumentCommitsSince(newCommitJson.version) // TODO: Implement
```
I don't know how many commit steps for a regular document, so I am not sure if this brings a storage burden for the server.

2. After 
```
const {docJson, commitJson} = applyCommitJSON(version, schema, doc, newCommits, newCommitJson)
```
a client receives the **right** `commitJson`, and rebases to the latest version. 
How comes "old and future commit versions"?

**Reply by ProTip** [reply]
Sorry for the late responses; I'm overseas through the end of this month.

> Well, I did not use prosemirror-collab before. Forgive me if I asked stupid questions

No worries; I just reference it as this plugin is built on the bones of `prosemirrelor-collab` and anticipates familiarity with it.


### 1.

Yes, but only for as far back as the oldest commit base version you want to accept. It would be unusual in my experience to accept very old commits, and then storage requirements are suprisingly modest for even 60k steps with something like postgres jsonb since it effectively deduplicates keys.

You could downres commits after a week or month(or whatever makes sense) if the storage becomes a major concern in the future.

### 2.

If the plugin state is on confirmed version 5 it will ignore commits from the authority that are not version 6(expected next version); https://github.com/stepwisehq/prosemirror-collab-commit/blob/33a6e55444bad759bfdbdc145826ae59f5393e79/src/collab-commit.ts#L229

This is to protect against applying out-of-order commits.

**Reply by etclub** [reply]
Sorry to interrupt you again.
But I still cannot create new sendable commits.

The client inputs "A" in the editor, and sends `sendableCommit(state)` to the server, something like:
```
 {
  version: 0,
  ref: '7x1n5dshnuvr',
  steps: [ { stepType: 'replace', from: 1, to: 1, slice: [Object] } ]
}
```
The server gets `commitJSON` (returned from `applyCommitJSON`), something like:
```
{
  version: 1,
  ref: '7x1n5dshnuvr',
  steps: [ { stepType: 'replace', from: 1, to: 1, slice: [Object] } ]
}
```
and take it as a response to the client. 

The client deals with the response:
```
const commitJson = await response.json();
const commit = Commit.FromJSON(state.schema, commitJson)
const tr = receiveCommitTransaction(state, commit)
state = state.apply(tr)                           // I also tried    dispatch(tr)       , but it does not help 
```

Now no matter what edits the client makes,
calling `sendableCommit(state)` still generates
```
{
  version: 0,
  ref: '7x1n5dshnuvr',
  steps: [ { stepType: 'replace', from: 1, to: 1, slice: [Object] } ]
}
```
which means the first commit is not confirmed, but I cannot see what is wrong.

**Reply by etclub** [reply]
I solved the problem.
I have to use the  **current** `state` of the editor.
